pkg/controller/cyclenoderequest/transitioner/util.go

```diff
@@ -295,31 +330,202 @@ func findProblemNodes(kubeNodes map[string]corev1.Node, nodeGroupInstances map[s
+		// more action on the node and go to the next one. This case can be fixed in the next
+		// run of the Pending phase.
+		if apierrors.IsNotFound(err) {
+			continue
+		}
+
+		// Otherwise error out to end the cycle. This includes if the cycling annotation is
+		// missing since there is no link between the original cloud provider instance and it's
+		// nodegroup so there's no way to re-attach it.
+		if err != nil {
+			return err
+		}
+
+		// AttachInstance does not error out if the instance does not exist so no need to handle
+		// it here. Error out on any error that can't be fixed by repeating the attempts to fix
+		// the instance state.
+		alreadyAttached, err := nodeGroups.AttachInstance(node.Spec.ProviderID, nodegroupName)
+		if err != nil && !alreadyAttached {
+			return err
+		}
 	}
 
-	// handle conflicts before complaining
-	if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		return t.rm.UpdateObject(t.cycleNodeRequest)
-	}); err != nil {
-		t.rm.Logger.Error(err, "unable to update cycleNodeRequest")
+	return nil
+}
+
+// deleteAnyOrphanedKubeNodes filters through the kube nodes without instances in the cloud provider
+// nodegroups which don't have an instance in the cloud provider at all. It removes the cycling
+// finalizer and deletes the node.
+func (t *CycleNodeRequestTransitioner) deleteAnyOrphanedKubeNodes(nodesNotInCloudProviderNodegroup map[string]corev1.Node) error {
+	var nodeProviderIDs []string
+
+	if len(nodesNotInCloudProviderNodegroup) == 0 {
+		return nil
 	}
 
-	// Notify that the cycling has transitioned phase
-	if t.rm.Notifier != nil {
-		if err := t.rm.Notifier.PhaseTransitioned(t.cycleNodeRequest); err != nil {
-			t.rm.Logger.Error(err, "Unable to post message to messaging provider", "phase", t.cycleNodeRequest.Status.Phase)
+	for _, node := range nodesNotInCloudProviderNodegroup {
+		nodeProviderIDs = append(nodeProviderIDs, node.Spec.ProviderID)
+	}
+
+	existingProviderIDs, err := t.rm.CloudProvider.InstancesExist(nodeProviderIDs)
+	if err != nil {
+		return errors.Wrap(err, "failed to check instances that exist from cloud provider")
+	}
+
+	// Find all the orphaned kube nodes from the set of nodes without matching instance in the
+	// cloud provider nodegroup.
+	for providerID, node := range nodesNotInCloudProviderNodegroup {
+		_, instanceExists := existingProviderIDs[providerID]
+
+		if instanceExists {
+			continue
+		}
+
+		// The kube node is now established to be orphaned. Check the finalizers on the node
+		// object and ensure that only the Cyclops finalizer exists on it. If another finalizer
+		// exists then it's from another controller and it should not be deleted.
+		containsNonCyclingFinalizer, err := t.rm.NodeContainsNonCyclingFinalizer(node.Name)
+		if err != nil {
+			return errors.Wrap(err,
+				fmt.Sprintf("failed to check if node %s contains a non-cycling finalizer", node.Name),
+			)
+		}
+
+		if containsNonCyclingFinalizer {
```

Comment:
I think if we detect a non cycling finalizer we should probably retry a few times to see if the other finalizer is acting on the object. This should only happen once the Node is actually being deleted though. We can start with this implementation, it seems fine.

Technically, are we misusing finalizers here? We could go all in on proper usage and have a webhook and everything to clean these up properly, but that also starts to impact cluster stability. In that case, should we juse use an annotation and no finalizer?

Reply:
> I think if we detect a non cycling finalizer we should probably retry a few times to see if the other finalizer is acting on the object. This should only happen once the Node is actually being deleted though. We can start with this implementation, it seems fine.

If we see issues I'm happy to go in a change it at a later point.

> Technically, are we misusing finalizers here? We could go all in on proper usage and have a webhook and everything to clean these up properly, but that also starts to impact cluster stability. In that case, should we juse use an annotation and no finalizer?

We're using a finalizer to prevent anything from prematurely deleting the node after it has been detached from the ASG. The finalizer is just there to allow cyclops to control the deletion process by drain pods as intended, etc. This is the intended use no?

Reply:
This is a valid use for a finalizer, agreed. Thanks for clarifying.